go/ct/spc/coordination_test.go

```diff
@@ -0,0 +1,186 @@
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+
+	if counter.Load() != int64(statesPerRule) {
+		t.Errorf("unexpected number of evaluated states: %d", counter.Load())
+	}
+}
+
+func TestCoordination_AllRulesAreEnumerated(t *testing.T) {
+	numJobs := runtime.NumCPU()
+	seed := 0
+	var counter atomic.Int64
+	opFunction := func(state *st.State) rlz.ConsumerResult {
+		counter.Add(1)
+		return rlz.ConsumeContinue
+	}
+	printFunction := func(time time.Duration, rate float64, current int64) {}
+
+	//var condition testCondition
+	rules := []rlz.Rule{}
+	testRuleFit := rlz.Rule{
+		Name:      "test_rule",
+		Condition: condition(true, 1),
+		Effect:    rlz.NoEffect(),
+	}
+	testRuleNoFit := rlz.Rule{
+		Name:      "test_rule",
+		Condition: condition(false, 1),
+		Effect:    rlz.NoEffect(),
+	}
+
+	for i := 0; i < rand.Intn(42); i++ {
+		rules = append(rules, testRuleFit)
+		rules = append(rules, testRuleNoFit)
+	}
+	numRules := len(rules)
+
+	err := ForEachState(rules, opFunction, printFunction, numJobs, uint64(seed), false)
+	if err != nil {
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+
+	if counter.Load() != int64(numRules)/2 {
+		t.Errorf("unexpected number of evaluated states: %d", counter.Load())
+	}
+
+	counter.Store(0)
+	err = ForEachState(rules, opFunction, printFunction, numJobs, uint64(seed), true)
+	if err != nil {
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+
+	if counter.Load() != int64(numRules) {
+		t.Errorf("wrong number of evaluated states in full mode: %d vs. %d", counter.Load(), numRules)
+	}
+}
+
+func TestCoordination_FilterRules(t *testing.T) {
+	filters := []string{"add", "sub", "mul", "copy", "call"}
+	for _, subString := range filters {
+		filter, err := regexp.Compile(subString)
+		if err != nil {
+			t.Error("regular expression not compilable")
+		}
+		rules := Spec.GetRules()
+		rules = FilterRules(rules, filter)
+
+		ruleNames := make([]string, 0, len(rules))
+		for _, rule := range rules {
+			ruleNames = append(ruleNames, rule.Name)
+		}
+
+		for _, name := range ruleNames {
+			if !strings.Contains(name, subString) {
+				t.Errorf("rules not filtered correctly %v", name)
+			}
+		}
+	}
+}
+
+func TestCoordination_EmptyRules(t *testing.T) {
+	numJobs := 1
+	seed := 0
+	fullMode := false
+	opFunction := func(state *st.State) rlz.ConsumerResult {
+		return rlz.ConsumeContinue
+	}
+	printFunction := func(time time.Duration, rate float64, current int64) {}
+	rules := []rlz.Rule{}
+	err := ForEachState(rules, opFunction, printFunction, numJobs, uint64(seed), fullMode)
+	if err != nil {
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+}
+
+func TestCoordination_RightNumberOfGoroutinesIsStarted(t *testing.T) {
+	numJobs := 4
+	seed := 0
+	fullMode := false
+	filter := regexp.MustCompile(".*")
+
+	opFunction := func(state *st.State) rlz.ConsumerResult {
+
+		// 3 goroutines (sweeper, scavenger and finalizer) are started by default
```

Comment:
Maybe instead of hard-coding the number of goroutines running outside of the code to be tested, query them before the `ForEachState` run and store it in a variable. This would make the test agnostic to the use of goroutines by the test framework or the runtime system.